include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -112,6 +113,183 @@ struct walk_n
+template <typename... _Ranges>
+struct walk_scalar_base
+{
+  private:
+    using _ValueTypes = std::tuple<oneapi::dpl::__internal::__value_t<_Ranges>...>;
+    constexpr static std::uint8_t __min_type_size = oneapi::dpl::__internal::__min_nested_type_size<_ValueTypes>::value;
+    constexpr static std::uint8_t __bytes_per_item = 16;
+
+  public:
+    constexpr static bool __can_vectorize = false;
+    // With no vectorization, the vector size is 1
+    constexpr static std::uint8_t __preferred_vector_size = 1;
+    // To achieve full bandwidth utilization, multiple iterations need to be processed by a work item
+    constexpr static std::uint8_t __preferred_iters_per_item =
+        __bytes_per_item / (__min_type_size * __preferred_vector_size);
+};
+
+template <typename _ExecutionPolicy, typename _F, typename _Range>
+struct walk1_vector_or_scalar : public walk_vector_or_scalar_base<_Range>
+{
+  private:
+    using __base_t = walk_vector_or_scalar_base<_Range>;
+    _F __f;
+    std::size_t __n;
+
+  public:
+    walk1_vector_or_scalar(_F __f, std::size_t __n) : __f(std::move(__f)), __n(__n) {}
+
+    template <typename _IsFull>
+    void
+    __vector_path_impl(_IsFull __is_full, const std::size_t __idx, _Range __rng) const
+    {
+        oneapi::dpl::__par_backend_hetero::__vector_walk<__base_t::__preferred_vector_size>{__n}(__is_full, __idx, __f,
+                                                                                                 __rng);
+    }
+
+    // _IsFull is ignored here. We assume that boundary checking has been already performed for this index.
+    template <typename _IsFull>
+    void
+    __scalar_path_impl(_IsFull, const std::size_t __idx, _Range __rng) const
+    {
+        __f(__rng[__idx]);
+    }
+
+    template <typename _IsFull>
+    void
+    operator()(_IsFull __is_full, const std::size_t __idx, _Range __rng) const
+    {
+        if constexpr (__base_t::__can_vectorize)
+            __vector_path_impl(__is_full, __idx, __rng);
+        else
+            __scalar_path_impl(__is_full, __idx, __rng);
+    }
+};
+
+template <typename _ExecutionPolicy, typename _F, typename _Range1, typename _Range2>
+struct walk2_vectors_or_scalars : public walk_vector_or_scalar_base<_Range1, _Range2>
+{
+  private:
+    using __base_t = walk_vector_or_scalar_base<_Range1, _Range2>;
+    _F __f;
+    std::size_t __n;
+
+  public:
+    walk2_vectors_or_scalars(_F __f, std::size_t __n) : __f(std::move(__f)), __n(__n) {}
+
+    template <typename _IsFull>
+    void
+    __vector_path_impl(_IsFull __is_full, const std::size_t __idx, _Range1 __rng1, _Range2 __rng2) const
+    {
+        using _ValueType1 = oneapi::dpl::__internal::__value_t<_Range1>;
+        _ValueType1 __rng1_vector[__base_t::__preferred_vector_size];
+        // 1. Load input into a vector
+        oneapi::dpl::__par_backend_hetero::__vector_load<__base_t::__preferred_vector_size>{__n}(
```

Comment:
Why it is so long?
Let me share my thoughts:
1) I would suggest making __vector_load as a method of the base class of `struct walk2_vectors_or_scalars`.
2) As far a a constructor had been already called, we have already had the initialized base class.

Taking into considerations the conclusions above we can write a short understandable code line, like this: 
`__vec_load(__idx, __rng1, __rng1_vector);`

Reply:
Instead of a function in the base class, I think an alias is more appropriate to simplify the name and still create the load / store objects on the device.

The problem with having it as a member of `walk2_vectors_or_scalars` is that it would require even more device copyable specializations as it would then be created on the host and copied to the device. I also assume there could potentially be higher launch overhead with this copy.

Reply:
I think I misunderstood the first point about turning `__vector_load` into a method. Is the suggestion here to pull the whole implementation as a member function into the base class? The reason I am against that is it would no longer be a generic utility which could be used elsewhere and constrain it to the parallel for implementation. The utilities introduced `__vector_load`, `__strided_loop`, etc.  I think are best fit to be decoupled from any algorithm. Since they are generic and enforce good access patterns, they may can be useful elsewhere when loading from global memory.